### PR TITLE
Improve scikit learn compat

### LINKED ISSUE #50: scikitlearn compatibility
Not all our estimators are compatible with scikit learn. For example, the usage of `**kwargs` in DecisionTreeClassifier:

```python
from sklearn.utils.estimator_checks import check_estimator
from skorecard.bucketers import DecisionTreeBucketer

check_estimator(DecisionTreeBucketer())
```

We should:

- Update unit tests to run `check_estimator` on all estimators
- Fix any issues that arise

I found out about this one because using `Skorecard()` inside `cross_validate` produced an error.

### PR DESCRIPTION
**Pull request by timvink**
This is a huge PR (~1000 lines of code across 30 files changes, excluding notebooks). It includes several breaking changes, warranting a `v2.0` upgrade imho. Commits in this PR address multiple issues, fixes quite some bugs, and overall ensure scikit-learn compatibility of different transformers and estimators in `skorecard`. 

It should fix #50, #47  and #48 

An overview of the changes:

## Changes

- `BucketingProcess` now does not enforce categorical features to be prebucketed, as it is not necessary.
- `skorecard.preprocessing.WoEEncoder` has now been deprecated in favor of `category_encoders.woe.WOEEncoder`, which is a new dependency.

These are breaking changes:

- `skorecard.Skorecard()` dropped support for `cat_columns` and `selected_features`. To make sure to bucket categoricals, supply a `bucketing` estimator.  If not supplied, categorical will be auto-detected. A new `variables` parameter will select features to be used in the `LogisticRegression`. A new `calculate_stats` controls if the `LogisticRegression` will calculate standards errors or not, and is set to False (previously this was always done).
- `skorecard.Skorecard()` has an updated default pipeline that will be used if no pipeline is specified. Tweaked some parameters in the bucketing and the detecting of categorical vs numerical columns.
- `skorecard.linear_model.LogisticRegression` now has a new `calculate_stats` parameter that control whether or not to calculate (expensive) statistics during `.fit()`. Defaults to False.
- `skorecard.bucketers.DecisionTreeBucketer()` has a new parameter `dt_kwargs`, with keyword arguments passed to the `DecisionTreeClassifier`. Previously kwargs were directly passed.
- `skorecard.bucketers.OptimalBucketer()` has a new parameter `ob_kwargs` with keyword arguments passed to `optimalbinning.OptimalBinning`. Also several new parameters exposed directly, see docs.
- Input validation on many classes have now been moved to the `fit()` method, and many internal class variables now have a trailing underscore (`_`) to signal they are determined during fit. This might cause errors to occur later in your skorecard pipeline. Reason for this is to properly support estimator cloning in scikit-learn, so you can use skorecard estimators in grid searches and cross validations.

## Added

- Improved error handling and input validation in many classes
- Added support for numpy matrixes, which are converted gracefully to `pd.DataFrame()`
- Expanded unit tests to use scikit learns suite of checks in [`check_estimator`](http://scikit-learn.org/stable/modules/generated/sklearn.utils.estimator_checks.check_estimator.html)

### CODE REVIEW

File: skorecard/bucketers/base_bucketer.py [hunk 1]
```diff
@@ -345,15 +368,30 @@ def transform(self, X: pd.DataFrame, y=None) -> pd.DataFrame:
             df (pd.DataFrame): dataset with transformed features
         """
         check_is_fitted(self)
-        X = self._is_dataframe(X)
+        X = ensure_dataframe(X)
         y = self._check_y(y)
+        if y is not None:
```
Comment (orchardbirds) [inline skorecard/bucketers/base_bucketer.py:373]: can we use sklearn's `check_X_y` for this?

File: skorecard/bucketers/bucketers.py [hunk 2]
```diff
@@ -67,7 +72,8 @@ def __init__(
                 This dictionary contains a name of a bucket (key) and an array of unique values that should be put
                 in that bucket.
                 When special values are passed, they are not considered in the fitting procedure.
-            variables_type: Type of the variables
+            variables_type: Passed to [optbinning.OptimalBinning](http://gnpalencia.org/optbinning/binning_binary.html):
+                Type of the variables. Much be either 'categorical' or 'numerical'.
```
Comment (orchardbirds) [inline skorecard/bucketers/bucketers.py:76]: "Must" not Much

File: skorecard/metrics/metrics.py [hunk 3]
```diff
@@ -48,6 +52,8 @@ def woe_1d(X, y, epsilon=0.00001):
     t = pd.concat([pos, neg], axis=1)
     t["woe"] = np.log(t["non_target"] / t["target"])
 
+    if t["woe"].isnull().any():
+        raise ValueError("Woe Calculation produced NaNs!")
```
Comment (orchardbirds) [inline skorecard/metrics/metrics.py:56]: Maybe add something here like "check your target distribution contains more than 1 class"

File: skorecard/skorecard.py [hunk 4]
```diff
@@ -156,114 +148,83 @@ def __repr__(self):
...
             bucketing_pipe.append(
                 OptimalBucketer(
-                    variables=self.cat_features,
+                    variables=cat_features,
                     variables_type="categorical",
                     max_n_bins=6,
                     min_bin_size=0.05,
                 )
             )
+        if len(num_features) == 0 and len(num_features) == 0:
```
Comment (orchardbirds) [inline skorecard/skorecard.py:176]: cat_features, you have num_features twice

File: skorecard/bucketers/bucketers.py (hunk 2 above)
Comment (timvink) [reply skorecard/bucketers/bucketers.py:76]: Good catch, will update

File: skorecard/metrics/metrics.py (hunk 3 above)
Comment (timvink) [reply skorecard/metrics/metrics.py:56]: Done.

File: skorecard/skorecard.py (hunk 4 above)
Comment (timvink) [reply skorecard/skorecard.py:176]: nice catch, fixed

**Pull request merged**